### ISSUE TITLE
Structure calculation mass from radius does not work with Aragog

### ISSUE THREAD

**Issue by lsoucasse** [FormingWorlds/PROTEUS#277]
With the new structure scheme, the mantle calculation from radius does not work because the gravity is no longer initialized and we need the gravity to compute the mantle mass with Aragog. You can actually see in the physical test results that both mantle mass and gravity are increasing during the first iterations (IC_INTERIOR = 1) up to a not converged value while it should be correct from the beginning.

I think for now we have to have the EOS function that uses Aragog in Proteus and invert these relationships to set either the radius from mass or the mass from radius.

**Comment by lsoucasse** [comment]
By the way, I don't understand why the structure calculation is within the time loop. As far as I understand all our interior models (dummy/Spider/Aragog) work with a fixed mesh/gravity/mantle_mass and we ignore mass variations due to the melting. Am I correct @nichollsh , @timlichtenberg ?

**Comment by lsoucasse** [comment]
I fully understand the mid-term goal of changing the structure with time.

But for now, it solves it 2/3 times in the init and also solves it again when we resume a simulation. I am curious to know how Spider behaves actually. Does it give constant (and consistent) results from first iteration?

**Comment by nichollsh** [comment]
I agree that it's definitely not optimal. For the default configuration (where structure is set by mass using SPIDER) I get this result:
![image](https://github.com/user-attachments/assets/61c51682-6015-43ba-a4a0-295c232a46b2)

**Comment by lsoucasse** [comment]
And what about the other way around, when you set the radius? I am just afraid spider does not compute the mass correctly, because it might need the gravity constant which is set to zero at the beginning now.

I think we need a `determine_interior_mass` function in the `solve_structure` function to get the mantle mass and gravity constant for a given radius (calling iteratively the mantle mass calculation with updated gravity constant).

**Comment by nichollsh** [comment]
Good point. I think there are actually two things here.

A) The Aragog case is obviously not right because of this `solve_g` flag being set to false. Setting it to true with Aragog causes the model to crash when I test it. We should definitely look into this.

B) Using SPIDER and setting the structure with the mass, `solve_g` is always true, so it will determine the gravity consistently with mass/radius. However, we should check that this is also fine when setting structure from radius, because as you say, the gravity needs to be solved self-consistently and this can only be done with a `determine_interior_mass` calculation.

**Comment by lsoucasse** [comment]
Yes I agree.

I will take care of Aragog.

For Spider, could you check the radius to mass case? I still have issues running petsc on my Mac. If you see an increasing gravity and mantle mass for the init iterations (starting from zero) then it does not work properly. In that case we can just set a warning flag for the time being.

**Comment by timlichtenberg** [comment]
Just adding to this thread here and to possibly simplifying some discussion. For using the EOS directly to derive the M-R relation, it will be a good enough first estimate to use the solid-only files. Typically it is expected that there shouldn't be a huge difference between melt- and solid-only models for the structure. Ultimately, of course we will want to move toward having a fully self-consistent structural and dynamical model, but a solid-only structural model is a reasonable guess for now.

**Comment by lsoucasse** [comment]
> @lsoucasse I have tested this in the radius -> mass case with SPIDER, and it seems to work okay.
> 
> I have also tested another method whereby it calls a new function `determine_interior_mass` which iterates SPIDER until the mass is unchanging between iterations. This yields the exact same result as the current implementation, so I think it's fine to stick with what we have.
> 
> This is how it looks in the helpfile:
> 
> ```
> Time         	separation   	period       	R_int        	M_int        	M_tot       
> 0.000000e+00 	1.495979e+11 	2.130945e+10 	5.701895e+06 	4.361010e+24 	4.361010e+24
> 0.000000e+00 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> 0.000000e+00 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> 0.000000e+00 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> 1.000000e+00 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> 2.000000e+00 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> 6.900000e+02 	1.495979e+11 	3.155816e+07 	5.701895e+06 	4.361010e+24 	4.361980e+24
> ```
> 
> This R_int corresponds to 0.9 R_earth, which is what I have specified in the configuration file.

Very good news, thanks.

**Comment by nichollsh** [comment]
There is a very small issue in the order in which the masses are calculated, however. It does not impact the results of the model because SPIDER/Aragog then recalculates the mass (and hence the helpfile values are fine) and PROTEUS the gravity.

When `set_by == 'radius_int'`, the `solve_structure()` function will output `M_int` and `gravity`. To do this, it pulls the mass from the configuration as an initial dummy value to get the first value for `gravity`. This is then overwritten by SPIDER inside the time-stepping loop, so it all works out okay. This means that between the structure and interior calculations being called, the value of `M_int` inside `hf_row` is actually incorrect. 

This could cause problems in the future, so I will make a small change to fix this potential problem. This can be done by ensuring that the dummy value for `M_int` is only ever applied in the first instance.

**Comment by lsoucasse** [comment]
OK, I have made some progress on this. Actually there is no reason why the determine_radius_interior would not work, it can go through aragog and get the mantle mass for a given radius, update the gravity field and loop until it converges.

The problem at the moment is that the mantle mass returned by aragog is wrong. For near correct values for core mass, gravity field, inner and outer radius I got a mantle mass of 9e24 kg instead of around 4e24kg. I don't know wheter we give the wrong parameter to the ESO (surface density, adiabatic bulk modulus) or if the mass calculation in Aragog is wrong.
